src/adt/queue.c

```diff
@@ -0,0 +1,81 @@
+#include "adt/queue.h"
+
+#include <string.h>
+#include <assert.h>
+
+struct queue {
+	/* Read and write offsets. When rd == wr, the queue is empty. rd
+	 * will always be <= wr, and wr <= capacity. */
+	size_t rd;
+	size_t wr;
+	size_t capacity;
+	void *q[1 /* capacity */];
```

Comment:
This would be a great place for a C99 flexible array member.  Maybe @katef will make a C99 exception?

Reply:
Yes, fine. We can have C99 per file :)